File: python/src/tools/feature_flags/create.py
import json
from typing import List, Optional
from pydantic import BaseModel
from src.tools.types import Context, Tool, ToolResult, TextContent
from src.schema.flags import FilterGroups
from src.lib.utils.api import get_project_base_url


class CreateFeatureFlagParams(BaseModel):
    name: str
    key: str
    description: str
    filters: FilterGroups
    active: bool
    tags: Optional[List[str]] = None


async def create_feature_flag_handler(context: Context, params: CreateFeatureFlagParams) -> ToolResult:
    project_id = await context.get_project_id()

    flag_result = await context.api.feature_flags(project_id).create(params)

    if not flag_result.success:
        raise Exception(f"Failed to create feature flag: {flag_result.error}")

    feature_flag_with_url = {
        **flag_result.data.model_dump(),
        "url": f"{get_project_base_url(project_id)}/feature_flags/{flag_result.data.id}"
    }

    return ToolResult(content=[TextContent(text=json.dumps(feature_flag_with_url))])


def create_feature_flag_tool() -> Tool[CreateFeatureFlagParams]:
    return Tool(
        name="create-feature-flag",
        description="""Creates a new feature flag in the project. Once you have created a feature flag, you should:
     - Ask the user if they want to add it to their codebase
     - Use the "search-docs" tool to find documentation on how to add feature flags to the codebase (search for the right language / framework)
     - Clarify where it should be added and then add it.
        """,
        schema=CreateFeatureFlagParams,
        handler=create_feature_flag_handler
    )

File: typescript/src/tools/surveys/create.ts
import { SurveyCreateSchema } from "@/schema/tool-inputs";
import { formatSurvey } from "@/tools/surveys/utils/survey-utils";
import { getToolDefinition } from "@/tools/toolDefinitions";
import type { Context, Tool } from "@/tools/types";
import type { z } from "zod";

const schema = SurveyCreateSchema;
type Params = z.infer<typeof schema>;

export const createHandler = async (context: Context, params: Params) => {
	const projectId = await context.stateManager.getProjectId();

	const surveyResult = await context.api.surveys({ projectId }).create({
		data: params,
	});

	if (!surveyResult.success) {
		throw new Error(`Failed to create survey: ${surveyResult.error.message}`);
	}

	const formattedSurvey = formatSurvey(surveyResult.data, context, projectId);

	return {
		content: [{ type: "text", text: JSON.stringify(formattedSurvey) }],
	};
};

const definition = getToolDefinition("survey-create");

const tool = (): Tool<typeof schema> => ({
	name: "survey-create",
	description: definition.description,
	schema,
	handler: createHandler,
	annotations: {
		destructiveHint: false,
		idempotentHint: false,
		openWorldHint: true,
		readOnlyHint: false,
	},
});

export default tool;
